src/test/java/org/folio/rest/api/InstanceStorageTest.java

```diff
@@ -1783,6 +1805,510 @@ public void cannotPostSynchronousBatchWithDuplicateId() throws Exception {
     }
   }
 
+  @Test
+  public void canGenerateInstanceHRIDWhenNotSupplied() throws Exception {
+    log.info("Starting canGenerateInstanceHRIDWhenNotSupplied");
+
+    final UUID id = UUID.randomUUID();
+    final JsonObject instanceToCreate = smallAngryPlanet(id);
+    instanceToCreate.remove("hrid");
+
+    setInstanceSequence(1);
+
+    createInstance(instanceToCreate);
+
+    final Response createdInstance = getById(id);
+
+    assertThat(createdInstance.getJson().getString("hrid"), is("in00000001"));
+
+    log.info("Finished canGenerateInstanceHRIDWhenNotSupplied");
+  }
+
+  @Test
+  public void canCreateInstanceWhenHRIDSupplied() throws Exception {
+    log.info("Starting canCreateInstanceWhenHRIDSupplied");
+
+    final String hrid = "testHRID";
+    final UUID id = UUID.randomUUID();
+    final JsonObject instanceToCreate = smallAngryPlanet(id);
+    instanceToCreate.put("hrid", hrid);
+
+    createInstance(instanceToCreate);
+
+    final Response createdInstance = getById(id);
+
+    assertThat(createdInstance.getJson().getString("hrid"), is(hrid));
+
+    log.info("Finished canCreateInstanceWhenHRIDSupplied");
+  }
+
+  @Test
+  public void cannotCreateInstanceWithDuplicateHRID() throws Exception {
+    log.info("Starting cannotCreateInstanceWithDuplicateHRID");
+
+    final UUID id = UUID.randomUUID();
+    final JsonObject instanceToCreate = smallAngryPlanet(id);
+    instanceToCreate.remove("hrid");
+
+    setInstanceSequence(1);
+
+    createInstance(instanceToCreate);
+
+    final Response createdInstance = getById(id);
+
+    assertThat(createdInstance.getJson().getString("hrid"), is("in00000001"));
+
+    final JsonObject instanceToCreateWithSameHRID = nod(UUID.randomUUID());
+    instanceToCreateWithSameHRID.put("hrid", "in00000001");
+
+    final CompletableFuture<Response> createCompleted = new CompletableFuture<>();
+
+    client.post(instancesStorageUrl(""), instanceToCreateWithSameHRID, TENANT_ID,
+      text(createCompleted));
+
+    final Response response = createCompleted.get(5, SECONDS);
+
+    assertThat(response.getStatusCode(), is(400));
+    assertThat(response.getBody(),
+        is("duplicate key value violates unique constraint \"instance_hrid_idx_unique\": Key (lower(f_unaccent(jsonb ->> 'hrid'::text)))=(in00000001) already exists."));
```

Comment:
Are we able to replace this message with a more human friendly message?

Reply:
Sure. I was using `PgExceptionUtil` here because I thought that might be the standard way we want to send back SQL errors so the UI can deal with them. I also wanted to avoid processing this string. It is a pretty gnarly message, but it is somewhat consistent across all 3 instance creation APIs, each which a different format of error message/structure. I'll look into making a more human readable message that will work across all 3 APIs.

Reply:
I agree that depending upon parsing database error text is not ideal, and potentially fiddly and brittle.

I think that is preferable to sending raw database errors to the client (or at the least, only those).

Reply:
@marcjohnson-kint the more I look into this, the more challenging this appears to be. For the single API call, it may not be so bad, but to make this work in the sync batch API, I need to either modify the RMB ValidationHelper or put some special case logic in StorageHelper (at least it is in this module), which is used by all the sync APIs. I can do it, but it will take some time. I think the same thing is true for the older batch API as the single instance API, the code is just entirely different.

Do you feel strongly about hiding the SQL for this PR? Maybe we just modify the single instance case and let the batch APIs return the SQL message? What do you think? I am concerned we are creating a special case for this one error and all other SQL errors are just errors with SQL messages. It'd be nice if we had a SQL to human readable message mapper, which is what I was hoping `PgExceptionUtil` would do.

Reply:
Not so strongly as to delay this work to that degree.

When I was introducing the early database constraints, prior to the exception until, I did some messy string matching.

I think this probably needs more thought than this individual work really allows for.